File: java/src/main/java/problems101_200/IntersectionOfTwoLinkedLists.java
package problems101_200;

import concepts.SinglyListNode;

enum IntersectionOfTwoLinkedLists {
    DEFAULT {
        @Override
        SinglyListNode getIntersectionNode(SinglyListNode headA, SinglyListNode headB) {
            SinglyListNode nodeA = headA;
            SinglyListNode nodeB = headB;
            while (nodeA != nodeB) {
                nodeA = nodeA != null ? nodeA.next : headB;
                nodeB = nodeB != null ? nodeB.next : headA;
            }
            return nodeA;
        }
    };

    abstract SinglyListNode getIntersectionNode(SinglyListNode headA, SinglyListNode headB);
}

File: kotlin/src/main/kotlin/problems101_200/IntersectionOfTwoLinkedLists.kt
package problems101_200

import concepts.SinglyListNode

enum class IntersectionOfTwoLinkedLists {
    DEFAULT {
        override fun getIntersectionNode(
            headA: SinglyListNode?,
            headB: SinglyListNode?,
        ): SinglyListNode? {
            var nodeA = headA
            var nodeB = headB
            while (nodeA !== nodeB) {
                nodeA = if (nodeA != null) nodeA.next else headB
                nodeB = if (nodeB != null) nodeB.next else headA
            }
            return nodeA
        }
    },
    ;

    abstract fun getIntersectionNode(
        headA: SinglyListNode?,
        headB: SinglyListNode?,
    ): SinglyListNode?
}
